PMax Assets: AssetField for constructing the layout for editing images and texts

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR implements a part of 📌 [Assets form and related components](https://github.com/woocommerce/google-listings-and-ads/issues/1787#fe-misc) in #1787.

AssetField component includes:

- Construct the basic layout like heading, subheading, help content, etc.
- Toggle the content
- Composite the help popover
- Display the number of issues with a Pill
- Disableable

### Screenshots:

<img width="632" alt="2023-01-16 17 18 26" src="https://user-images.githubusercontent.com/17420811/212666895-d3e10031-6957-48c0-9d2d-0ea4d54616b5.png">

### Detailed test instructions:

1. Go to the assets step of the campaign creation page.
2. Use the AssetField demo to test.

### 📌 Dev notes

- [x] Revert fc5c54c46d4eb096ba0e396147ffdb78633530fc before merging.

### Changelog entry

eason9487 committed:
Limit the application scope of the header style to Section component only

eason9487 committed:
Change the id prop of HelpPopover to be optional

eason9487 committed:
Make the className and iconSize props of HelpPopover configurable and tweak its style

eason9487 committed:
Add a new prop `disabled` to HelpPopover

eason9487 committed:
Extract the add button from ImagesSelector

eason9487 committed:
Add a new component AssetField for constructing the layout and the common parts for editing images and texts

eason9487 committed:
Add tests for AssetField component

eason9487 committed:
Add demo for AssetField

puntope commented on js/src/components/help-popover/index.js:
```diff
@@ -20,30 +22,45 @@ import './index.scss';
-			id,
-		} );
+
+		if ( id ) {
+			recordEvent( 'gla_tooltip_viewed', { id } );
+		}
```
❓  Why are we making this optional now? It's a requirement ?

eason9487 replied:
Whether to do event tracking here is not yet decided. With the current implementation, `AssetField` uses `HelpPopover` but doesn't set `id` prop, so it was changed to optional.

puntope commented on js/src/components/paid-ads/asset-group/asset-field.js:
```diff
@@ -0,0 +1,130 @@
+import HelpPopover from '.~/components/help-popover';
+import './asset-field.scss';
+
+/**
+ * @typedef {Object} AssetFieldHandler
+ * @property {() => void} scrollIntoComponent Scroll to the nearest alignable position to show this component in the visible area.
+ */
```
💅  Following previous conventions. Shouldn't be this inside a file named types ?

eason9487 replied:
I thought the types.js file is mainly for keeping type definitions that could/would be imported across multiple files. `AssetFieldHandler` would probably only be used here.

puntope commented on js/src/components/paid-ads/asset-group/asset-field.js:
```diff
@@ -0,0 +1,130 @@
+ * @param {JSX.Element} props.subheading Subheading.
+ * @param {JSX.Element} props.help Help content to be shown after clicking on the help icon.
+ * @param {number} props.numOfIssues The number of issues used to label on UI. It only labels when the number is greater than 0.
+ * @param {boolean} [props.initialExpanded=false] Whether the UI to initial with expanded content.
```
💅  WDYT of rephrasing this?

`Whether the UI is initialised expanded.`

puntope replied:
Or

`If true, the UI is initialised in expanded state`

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1853/commits/4c723550d6d9be64d8706cf6ca255a68c0858920.

puntope commented on js/src/components/paid-ads/asset-group/asset-field.js:
```diff
@@ -0,0 +1,130 @@
+ * @param {JSX.Element} props.heading Heading.
+ * @param {JSX.Element} props.subheading Subheading.
+ * @param {JSX.Element} props.help Help content to be shown after clicking on the help icon.
+ * @param {number} props.numOfIssues The number of issues used to label on UI. It only labels when the number is greater than 0.
```
💅  `The number of issues used in the UI labels. It only shows the labels when the number is greater than 0.`

puntope replied:
Another option would be just to explain the implicit meaning:

`Number of issues in an asset field.  It shows labels in the UI when the number is greater than 0`

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1853/commits/4c723550d6d9be64d8706cf6ca255a68c0858920.

puntope commented on js/src/components/paid-ads/asset-group/asset-field.js:
```diff
@@ -0,0 +1,130 @@
+	ref
+) {
+	const containerRef = useRef();
+	const [ isExpanded, setExpanded ] = useState( initialExpanded );
```
💅  I personally would rename this as `[ isExpanded, setIsExpanded ]` 

or

`[ expanded, setExpanded ]`

eason9487 replied:
In this codebase, there are many of the same namings but they didn't seem to have caused confusion before. I would like to keep it as is if it still doesn't cause confusion.

puntope replied:
I'm not aware of those and this is not a blocker. But for me the standard React way of calling it [property, setProperty] is more convenient. And here ,especially, since there are several expanded related variables (initalExpanded, shouldExpanded and isExpanded).

eason9487 replied:
Thanks for the explanation. Renamed in https://github.com/woocommerce/google-listings-and-ads/pull/1853/commits/a33c984fc67205c07344923348205a3991e4120c.

To clarify, I'm not pretty sure if there is a standard React way to name the returned pair of `useState`. My understanding is that it's only a convention but not a standard, and WP and WC also have some cases in the same naming way. However, following the most popular convention on the naming of `useState` is still better.

puntope commented on js/src/components/paid-ads/asset-group/asset-field.js:
```diff
@@ -0,0 +1,130 @@
+		disabled ? 'gla-asset-field--is-disabled' : false
+	);
+
+	const shouldExpanded = isExpanded && ! disabled;
```
💅  Another renaming opinion.

Maybe `shouldExpand`, `shouldBeExpanded`, `isVisible` or `isExpanded` (this last one if we replace with the second option in my comment here https://github.com/woocommerce/google-listings-and-ads/pull/1853/files#r1072059197)

eason9487 replied:
Renamed in https://github.com/woocommerce/google-listings-and-ads/pull/1853/commits/4211d7cceb0d26d08d51c24c93a1124d1796e309.

puntope commented on js/src/components/paid-ads/asset-group/asset-field.js:
```diff
@@ -0,0 +1,130 @@
+	};
+
+	const issuePillText = sprintf(
+		// translators: %d: number of issues in a asset field.
```
Typo `an asset field.`

eason9487 replied:
Thanks! Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/1853/commits/4c723550d6d9be64d8706cf6ca255a68c0858920.

puntope commented on js/src/components/paid-ads/asset-group/asset-field.js:
```diff
@@ -0,0 +1,130 @@
+
+	const isReducedMotion = useReducedMotion();
+
+	useImperativeHandle( ref, () => ( {
+		scrollIntoComponent() {
+			containerRef.current.scrollIntoView( {
+				behavior: isReducedMotion ? 'auto' : 'smooth',
+				inline: 'nearest',
+				block: 'nearest',
+			} );
+		},
+	} ) );
```
❓  Maybe beyond my React knowledge. Could you explain to me what this code does? I don't see where ref is coming from .In my tests it's always `null`

eason9487 replied:
Sorry, I forgot to add a demo for it. Please check this commit https://github.com/woocommerce/google-listings-and-ads/pull/1853/commits/fc5c54c46d4eb096ba0e396147ffdb78633530fc.


https://user-images.githubusercontent.com/17420811/213081367-ccec1a97-eef1-4df6-81b1-a3f08ce70e5b.mp4

puntope reviewed (commented):
Howdy @eason9487 

I checked the behaviour and is looking good. Also I checked the code.

I have some questions and suggestions added in the review. Thanks

eason9487 committed:
Adjust JSDoc and fix a typo for AssetField.

Address:
- https://github.com/woocommerce/google-listings-and-ads/pull/1853#discussion_r1072053340
- https://github.com/woocommerce/google-listings-and-ads/pull/1853#discussion_r1072056812
- https://github.com/woocommerce/google-listings-and-ads/pull/1853#discussion_r1072062794

eason9487 committed:
Add a demo for scrollIntoComponent.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1853#discussion_r1072077725

eason9487 committed:
Rename shouldExpanded to shouldExpand for AssetField.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1853#discussion_r1072060560

eason9487 commented:
Hi @puntope, thanks for the review. This PR is ready for the second round.

puntope reviewed (approved):
✅  LGTM

Thanks for the adjustments. Looking good. I still disagree with the comment here

https://github.com/woocommerce/google-listings-and-ads/pull/1853/files#r1072059197

But it's not a blocker. So this is approved

eason9487 committed:
Rename isExpanded to expanded for AssetField.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1853#discussion_r1072059197

eason9487 committed:
Revert "Add a demo for scrollIntoComponent."

This reverts commit fc5c54c46d4eb096ba0e396147ffdb78633530fc.
